File: dsa-problems/Trees/solutions/cpp/TRE-017-lab-tree-right-view-skips.cpp
#include <iostream>
#include <vector>
#include <map>
#include <algorithm>

using namespace std;

class Solution {
    int maxDepth = -1;
    void dfs(int u, int depth, const vector<int>& values, const vector<int>& left, const vector<int>& right, map<int, int>& view) {
        if (u == -1) return;

        maxDepth = max(maxDepth, depth);

        if (values[u] >= 0 && view.find(depth) == view.end()) {
            view[depth] = values[u];
        }

        dfs(right[u], depth + 1, values, left, right, view);
        dfs(left[u], depth + 1, values, left, right, view);
    }

public:
    vector<int> rightViewWithSkips(int n, const vector<int>& values,
                                   const vector<int>& left, const vector<int>& right) {
        if (n == 0) return {};

        map<int, int> view;
        maxDepth = -1;
        dfs(0, 0, values, left, right, view);

        vector<int> result;
        for (int d = 0; d <= maxDepth; d++) {
            if (view.count(d)) {
                result.push_back(view[d]);
            }
        }
        return result;
    }
};

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    int n;
    if (!(cin >> n)) return 0;
    vector<int> values(n), left(n), right(n);
    for (int i = 0; i < n; i++) {
        cin >> values[i] >> left[i] >> right[i];
    }

    Solution solution;
    vector<int> ans = solution.rightViewWithSkips(n, values, left, right);
    for (int i = 0; i < (int)ans.size(); i++) {
        if (i) cout << ' ';
        cout << ans[i];
    }
    return 0;
}

File: dsa-problems/Trees/solutions/java/TRE-017-lab-tree-right-view-skips.java
import java.util.*;

class Solution {
    public List<Integer> rightViewWithSkips(int n, int[] values, int[] left, int[] right) {
        if (n == 0) return new ArrayList<>();
        
        Map<Integer, Integer> view = new HashMap<>();
        int maxDepth = dfs(0, 0, values, left, right, view);
        
        List<Integer> result = new ArrayList<>();
        for (int d = 0; d <= maxDepth; d++) {
            if (view.containsKey(d)) {
                result.add(view.get(d));
            }
        }
        return result;
    }

    private int dfs(int u, int depth, int[] values, int[] left, int[] right, Map<Integer, Integer> view) {
        if (u == -1) return -1;

        // If valid and first time seeing this depth (since we go Right -> Left)
        if (values[u] >= 0 && !view.containsKey(depth)) {
            view.put(depth, values[u]);
        }

        int d1 = dfs(right[u], depth + 1, values, left, right, view);
        int d2 = dfs(left[u], depth + 1, values, left, right, view);
        
        return Math.max(depth, Math.max(d1, d2));
    }
}

class Main {
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        if (!sc.hasNextInt()) return;
        int n = sc.nextInt();
        int[] values = new int[n];
        int[] left = new int[n];
        int[] right = new int[n];
        for (int i = 0; i < n; i++) {
            values[i] = sc.nextInt();
            left[i] = sc.nextInt();
            right[i] = sc.nextInt();
        }

        Solution solution = new Solution();
        List<Integer> ans = solution.rightViewWithSkips(n, values, left, right);
        StringBuilder sb = new StringBuilder();
        for (int i = 0; i < ans.size(); i++) {
            if (i > 0) sb.append(' ');
            sb.append(ans.get(i));
        }
        System.out.println(sb.toString());
        sc.close();
    }
}
